src/gorilla.c

```diff
@@ -0,0 +1,451 @@
+// e.g. for n=6, range is [-32, 31]
+
+static bool Bin_InRange(int64_t x, u_int8_t nbits) {
+    return x >= Bin_MinVal(nbits) && x <= Bin_MaxVal(nbits);
+}
+
+static inline binary_t *Bins_bitbin(u_int64_t *bins, globalbit_t bit) {
+    return &bins[bit / BINW];
+}
+
+static inline bool Bins_bitoff(u_int64_t *bins, globalbit_t bit) {
+    return !(bins[bit / BINW] & BIT(localbit(bit)));
+}
+
+static inline bool Bins_biton(u_int64_t *bins, globalbit_t bit) {
+    return !Bins_bitoff(bins, bit);
+}
+
+// Append `dataLen` bits from `data` into `bins` at bit position `bit`
+static void appendBits(binary_t *bins, globalbit_t *bit, binary_t data, u_int8_t dataLen) {
+    binary_t *bin_it = Bins_bitbin(bins, *bit);
+    localbit_t lbit = localbit(*bit);
+    localbit_t available = BINW - lbit;
+
+    if (available >= dataLen) {
+        *bin_it |= LSB(data, dataLen) << lbit;
+    } else {
+        *bin_it |= data << lbit;
+        *(++bin_it) |= LSB(data, dataLen) >> available;
+    }
+    *bit += dataLen;
+}
+
+// Read `dataLen` bits from `bins` at position `bit`
+static binary_t readBits(binary_t *bins, globalbit_t *bit, u_int8_t dataLen) {
+    binary_t *bin_it = Bins_bitbin(bins, *bit);
+    localbit_t lbit = localbit(*bit);
+    localbit_t available = BINW - lbit;
+
+    binary_t bin = 0;
+    if (available >= dataLen) {
+        bin = LSB(*bin_it >> lbit, dataLen);
+    } else {
+        u_int8_t left = dataLen - available;
+        bin = LSB(*bin_it >> lbit, available);
+        bin |= LSB(*++bin_it, left) << available;
+    }
+    *bit += dataLen;
+    return bin;
+}
+
+static bool isSpaceAvailable(CompressedChunk *chunk, u_int8_t size) {
+    u_int64_t available = (chunk->size * 8) - chunk->idx;
+    return size <= available;
+}
+
+/***************************** APPEND ********************************/
+static ChunkResult appendTS(CompressedChunk *chunk, timestamp_t timestamp) {
+    assert(timestamp >= chunk->prevTimestamp);
+    timestamp_t curDelta = timestamp - chunk->prevTimestamp;
+    
+    union64bits doubleDelta;
+    doubleDelta.i = curDelta - chunk->prevTimestampDelta;
+/* 
+ * Before any insertion the code `CHECKSPACE` ensures there is enough space to
+ * encode timestamp and one additional bit which the minimum to encode the value.
+ * This is why we have `+ 1` in `CHECKSPACE`.
+ * 
+ * If doubleDelta == 0, 1 bit of value 0 is inserted.
+ * 
+ * Else, `Bin_InRange` checks for the minimal number of bits required to represent
+ * `doubleDelta`, the delta of deltas between current and previous timestamps.
+ * Then two values are being inserted. 
+   * The first value is, encoding for the lowest number of bits for which
+     `Bin_InRange` returns `true`.
+   * The second value is a compressed representation of the value with the `length`
+     encoded by the first value. Compression is done using `int2bin`.
+ */
+    u_int64_t *bins = chunk->data;
+    u_int64_t *bit = &chunk->idx;
```

Comment:
@ashtul I think you want to use `globalbit_t` here, im not sure how valubale it is but it will be consistent with `appendV`